mrmustard/lab_dev/circuit_components_utils/branch_and_bound.py

```diff
@@ -0,0 +1,444 @@
+    the grandchildren level including leaf nodes whether they are
+    children or grandchildren.
+
+    Args:
+        graph (Graph): A graph.
+        cost_bound (int): The maximum cost of the grandchildren
+
+    Returns:
+        set[Graph]: The set of grandchildren below the cost bound.
+    """
+    grandchildren_set = set()
+    for child in children(graph, cost_bound):
+        if child.number_of_edges() == 0:
+            grandchildren_set.add(child)
+            continue
+        for grandchild in children(child, cost_bound):
+            grandchildren_set.add(grandchild)
+    return grandchildren_set
+
+
+def assign_costs(graph: Graph, debug: int = 0) -> None:
+    r"""
+    Assigns to each edge in the graph the cost of contracting the two nodes it connects.
+    """
+    for edge in graph.edges:
+        A = graph.nodes[edge[0]]["component"]
+        B = graph.nodes[edge[1]]["component"]
+        graph.edges[edge]["cost"] = A.contraction_cost(B)
+        if debug > 0:
+            print(
+                f"cost of edge {edge}: {A.repr}|{A.shape} x {B.repr}|{B.shape} = {graph.edges[edge]['cost']}"
+            )
+
+
+def random_solution(graph: Graph) -> Graph:
+    r"""
+    Returns a randomly contracted graph.
+    """
+    while graph.number_of_edges() > 0:
+        edge = random.choice(list(graph.edges))
+        graph = contract(graph, edge)
+    return graph
+
+
+def reduce_first(graph: Graph, code: str) -> tuple[Graph, Edge | bool]:
+    r"""
+    Reduces the first pair of nodes that match the pattern in the code.
+    The first number and letter describe a node with that number of
+    edges and that repr (B for Bargmann, F for Fock), and the last letter
+    describes the repr of the second node.
+    For example 1BB means we will contract the first occurrence of a node
+    that has one edge (a leaf) connected to a node of repr B with an arbitrary
+    number of edges.
+    We typically use codes like 1BB, 2BB, 1FF, 2FF by default because they are
+    safe, and codes like 1BF, 1FB optionally as they are not always the best choice.
+    """
+    n, tA, tB = code
+    for node in graph.nodes:
+        if int(n) == graph.degree(node):
+            for edge in list(graph.out_edges(node)) + list(graph.in_edges(node)):
+                A = graph.nodes[edge[0]]["component"]
+                B = graph.nodes[edge[1]]["component"]
+                if A.repr[0] == tA and B.repr[0] == tB:
+                    graph = contract(graph, edge)
+                    return graph, edge
+    return graph, False
+
+
+def heuristic(graph: Graph, code: str, verbose: bool) -> Graph:
+    r"""
+    Simplifies the graph by contracting all pairs of nodes that match the given pattern.
+    """
+    edge = True
+    while edge:
+        graph, edge = reduce_first(graph, code)
+        if edge and verbose:
+            print(f"{code} edge: {edge} | tot cost: {graph.cost}")
+    return graph
+
+
+def optimal_contraction(
```

Comment:
Actually should this file be broken down into a classes file and method file?

Reply:
You have a point, but it is also nice to have everything defined locally and avoid jumping around. I added two comment blocks for clarity.